### ISSUE TITLE
Keep include files location to avoid breaking builds

### ISSUE THREAD

**Issue by derpda** [laurikari/tre#92]
Glad to see that there is activity on this repo, and thank you @trushworth for doing work here.
However, as a regular user it would be nice for the header locations not to change every couple weeks as it breaks builds.

I realize I should probably fix the git commit that we use for our builds (and I will do that now that it has become an issue), but just in general it would be better to leave the default behavior intact.
The fact that there are no set releases makes this a bit harder to deal with.

If there is an installed version of `tre` in the system, it should be up to the user of the library to make sure their include/link paths are set correctly to pick the version they desire, and not up to the library itself.
As such, from my limited understanding of the library it seems to me that the changes of 32b2d6b4ef61bf5c900560f18d283c80e0fa4952 should be reverted.

**Comment by trushworth** [comment]
Hi,

It's good to know someone is using it besides me :).

Thanks for bringing this up, it is clear that even if I don't find a
better way to do the includes I need to put in a better explanation
of why I did what I did to them.  See below for the start of that.

On Mon, Jul 10, 2023 at 06:17:51PM -0700, derpda wrote:
> Glad to see that there is activity on this repo, and thank you @trushworth for doing work here.
> However, as a regular user it would be nice for the header locations not to change every couple weeks as it breaks builds.

I didn't really want to move them, but I was running into an issue with
different platforms and compilers where I'd pick up an installed header
instead of the build configured header.  The second time it happened and
I spent a couple of hours looking for the reason I said "enough" and
modified the source so that it _couldn't_ pick up an installed header.

The install location hasn't changed at all though.  It is still
   ${PREFIX}/include/tre
> 
> I realize I should probably fix the git commit that we use for our builds (and I will do that now that it has become an issue), but just in general it would be better to leave the default behavior intact.

Are you building TRE as a sub-component of a larger build, or building
your stuff against an installed TRE library?

> The fact that there are no set releases makes this a bit harder to deal with.

Nothing's going to happen for the next week, but after that I'm going to
go after a couple of the issues that are definitely bugs.  Once I'm done
with issues that are clearly bugs I'll bump the version number and look
at a release.
> 
> If there is an installed version of `tre` in the system, it should be up to the user of the library to make sure their include/link paths are set correctly to pick the version they desire, and not up to the library itself.

And it is, I think.  I'm not positive how the install location gets set
in the autotools though.  It's building the library from source that has
to get a specific header, different from anything that might have been
installed at some earlier time.  The installed header
"include/tre/tre-config.h" is different for different build
configurations (e.g. if you are using wide characters or not).  I had
installed a wide character build to test the install, but was getting
build failures with the python extension when I configured to build
without wide characters (and confusingly enough, only on some systems).

> As such, from my limited understanding of the library it seems to me that the changes of 32b2d6b4ef61bf5c900560f18d283c80e0fa4952 should be reverted.

I'll look for a better way than what I've done, but I don't really want
to leave it up to whichever compiler is being used, and what order that
compiler might search include directories in.  The changes in
32b2d6b4... are ugly, but compiler and platform independent, and
guarantee the library, agrep, and the python extension all get built
with the same headers from configuration.

The issue is probably pretty rare if you are building only one
configuration on one platform, because you are probably choosing the
same configuration each time you do it (so any installed headers are
the same as your build headers).  I'm building and testing 24
different configurations on both Linux and FreeBSD, and 12 on MacOS.

If you have a recent python handy, try the "waf" build if you want to
see what I do most of the time.

And thanks again for letting me know about this.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/laurikari/tre/issues/92
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

Cheers,
--
Tom Rushworth

**Comment by derpda** [comment]
Hey thanks for getting back to me so quickly!
Fair enough, with your explanation I can see what problem you were trying to solve.
If I was still a student I would delightfully forgo my duties and write up a CMake build for TRE... :smiling_face_with_tear: 

>Are you building TRE as a sub-component of a larger build, or building
your stuff against an installed TRE library?

I am using CMake to build a larger project, and during the configure step I build TRE (with some CMake black magic...), install it and use pkg-config to include it as a dependency. (I can go into more details, but maybe that is not relevant here).
Basically I call `./configure --prefix <my_prefix>`, `make` and then `make install`.
So I guess its a mix of both? However, if I don't define `USE_INSTALLED_TRE_H` with a compiler flag when using the (locally) installed TRE, I get include errors as it looks for headers in `../local_includes/`.

Best,
Peter

**Comment by trushworth** [comment]
Hi Peter,

On Mon, Jul 10, 2023 at 09:48:03PM -0700, derpda wrote:
> Hey thanks for getting back to me so quickly!

Not a problem, but I'm off for a few days starting about 2 hours from
now, so things will be quiet after this :).

> Fair enough, with your explanation I can see what problem you were trying to solve.
> If I was still a student I would delightfully forgo my duties and write up a CMake build for TRE... :smiling_face_with_tear: 

Well, I'm recently retired, and I do want to dig into this :).
> 
> >Are you building TRE as a sub-component of a larger build, or building
> your stuff against an installed TRE library?
> 
> I am using CMake to build a larger project, and during the configure step I build TRE (with some CMake black magic...), install it and use pkg-config to include it as a dependency. (I can go into more details, but maybe that is not relevant here).
> Basically I call `./configure --prefix <my_prefix>`, `make` and then `make install`.
> So I guess its a mix of both? However, if I don't define `USE_INSTALLED_TRE_H` with a compiler flag when using the (locally) installed TRE, I get include errors as it looks for headers in `../local_includes/`.

I'll see if I can reproduce this when I get back.  I suspect the problem
is that agrep includes "../local_includes/regex.h" which then goes on to
include the "../local_includes/tre.h" unless that flag is set.  I want to keep 
the USE_INSTALLED_TRE_H local to agrep and the python extension, not in
the headers themselves.

Hmm, another temporary route if you are not using agrep is to give the
configure step the flags to avoid building it.  That's not ideal, but it
might do for a couple of weeks if the USE_INSTALLED_TRE_H isn't what you
want.

Anyway I can't look at it now, but will when I get back.
> 
> Best,
> Peter
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/laurikari/tre/issues/92#issuecomment-1630110127
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

Cheers,
--
Tom Rushworth

**Comment by derpda** [comment]
Thanks for getting back to me, enjoy your days off! Looking forward to solving this with you.

**Comment by trushworth** [comment]
Hi derpda,

I thought about this while travelling, and realized that I got the responsibility for the controlling flag backwards.

The build of the TRE library, agrep, and the python extension are the places that should be setting any extra flags to force the local_includes.  External users of an installed TRE library should NOT have to set flags.

What I've done is to flip the sense of the "USE_INSTALLED_TRE_H" flag and rename it to "USE_LOCAL_TRE_H", then added definitions of the new name to all the local build methods.  This now builds and tests for both autoconf and waf on both Linux and FreeBSD.

There is no more "USE_INSTALLED_TRE_H" flag, but leaving it defined should have no effect.  Once the TRE headers are installed they should be usable by TRE library users without needing any extra flags.

Please let me know if I've shot off any more of my own toes with this :).  If this is OK, I'll close the issue.

Thanks again for pointing this problem out!

**Comment by derpda** [comment]
Build works fine again without setting `USE_INSTALLED_TRE_H`, thank you for the fix!
I think this is probably the right way to handle the issue of local vs installed TRE versions.

I'll go ahead and close the issue with this comment.
